// simple-angular-calendar/esm5/lib/models/views/moth-view.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { DayOfWeek } from '../day-of-week';
import { DateMath } from '../dates/date-math';
/**
 * Creates a view of a month.
 */
var /**
 * Creates a view of a month.
 */
MonthView = /** @class */ (function () {
    /**
     * Creates a new instance of MonthView.
     * @param date Date of the month.
     */
    function MonthView(date) {
        this.date = date;
        this.WEEKS_PER_MONTH = 5;
        this.DAYS_PER_WEEK = 7;
    }
    /**
     * Creates the grid corresponding to the month specified
     * by the date.
     * @param completeHoles Indicates if the beginning and end
     * of the grid should be completed with the days of the
     * adjacent months.
     * @param weekStart Day of week to be considered the beginning
     * of the week.
     */
    /**
     * Creates the grid corresponding to the month specified
     * by the date.
     * @param {?=} completeHoles Indicates if the beginning and end
     * of the grid should be completed with the days of the
     * adjacent months.
     * @param {?=} weekStart Day of week to be considered the beginning
     * of the week.
     * @return {?}
     */
    MonthView.prototype.createView = /**
     * Creates the grid corresponding to the month specified
     * by the date.
     * @param {?=} completeHoles Indicates if the beginning and end
     * of the grid should be completed with the days of the
     * adjacent months.
     * @param {?=} weekStart Day of week to be considered the beginning
     * of the week.
     * @return {?}
     */
    function (completeHoles, weekStart) {
        if (completeHoles === void 0) { completeHoles = false; }
        if (weekStart === void 0) { weekStart = DayOfWeek.Monday; }
        /** @type {?} */
        var currentMonthGrid = this.createGrid(this.date, weekStart);
        if (completeHoles) {
            /** @type {?} */
            var needsToBeFilledFromBeginning = !!!currentMonthGrid[0][0];
            if (needsToBeFilledFromBeginning) {
                /** @type {?} */
                var lastMonth = DateMath.substractMonths(this.date, 1);
                /** @type {?} */
                var lastDayOfLastMonth = DateMath.getLastDayOfMonth(lastMonth).getDate();
                /** @type {?} */
                var firstDayOfCurrentMonth = DateMath.getFirstDayOfMonth(this.date);
                /** @type {?} */
                var daysToPickFromLastMonth = firstDayOfCurrentMonth.getDay() - weekStart.valueOf();
                for (var i = 0; i < daysToPickFromLastMonth; i++) {
                    /** @type {?} */
                    var currentDayOfMonth = lastDayOfLastMonth - (daysToPickFromLastMonth - (i + 1));
                    /** @type {?} */
                    var currentDate = new Date(lastMonth.getFullYear(), lastMonth.getMonth(), currentDayOfMonth);
                    currentMonthGrid[0][i] = {
                        isToday: false,
                        isSelected: false,
                        day: currentDayOfMonth,
                        date: currentDate
                    };
                }
            }
            /** @type {?} */
            var fillingDays = 1;
            for (var week = 0; week < this.WEEKS_PER_MONTH; week++) {
                for (var dayOfWeek = 0; dayOfWeek < this.DAYS_PER_WEEK; dayOfWeek++) {
                    if (!!!currentMonthGrid[week][dayOfWeek]) {
                        /** @type {?} */
                        var fillingDates = new Date(this.date.getFullYear(), this.date.getMonth(), fillingDays++);
                        currentMonthGrid[week][dayOfWeek] = {
                            isToday: false,
                            isSelected: false,
                            day: fillingDays,
                            date: fillingDates
                        };
                    }
                }
            }
        }
        return currentMonthGrid;
    };
    /**
     * Creates the grid corresponding to the month specified
     * by the date.
     * @param {?} date Date to use.
     * @param {?=} weekStart Day of week which will be considered
     * the beginning of the week.
     * @return {?}
     */
    MonthView.prototype.createGrid = /**
     * Creates the grid corresponding to the month specified
     * by the date.
     * @param {?} date Date to use.
     * @param {?=} weekStart Day of week which will be considered
     * the beginning of the week.
     * @return {?}
     */
    function (date, weekStart) {
        if (weekStart === void 0) { weekStart = DayOfWeek.Monday; }
        /** @type {?} */
        var grid = this.initGrid();
        /** @type {?} */
        var firstDayOfWeek = DateMath.getFirstDayOfMonth(date).getDay() - weekStart.valueOf();
        /** @type {?} */
        var lastDay = DateMath.getLastDayOfMonth(date).getDate();
        /** @type {?} */
        var today = new Date();
        /** @type {?} */
        var dayOfWeek = firstDayOfWeek;
        /** @type {?} */
        var currentDayOfMonth = 1;
        for (var week = 0; week < this.WEEKS_PER_MONTH; week++) {
            for (; dayOfWeek < this.DAYS_PER_WEEK && currentDayOfMonth <= lastDay; dayOfWeek++) {
                /** @type {?} */
                var todaySameYear = today.getFullYear() === this.date.getFullYear();
                /** @type {?} */
                var todaySameMonth = today.getMonth() === this.date.getMonth();
                /** @type {?} */
                var todaySameDay = today.getDate() === currentDayOfMonth;
                /** @type {?} */
                var currentDate = new Date(this.date.getFullYear(), this.date.getMonth(), currentDayOfMonth);
                grid[week][dayOfWeek] = {
                    isToday: todaySameYear && todaySameMonth && todaySameDay,
                    isSelected: this.date.getDate() === currentDayOfMonth,
                    day: currentDayOfMonth,
                    date: currentDate
                };
                currentDayOfMonth++;
            }
            dayOfWeek = 0;
        }
        return grid;
    };
    /**
     * Initializes the grid that contains
     * the info about the "shape" of the
     * month.
     * @return {?}
     */
    MonthView.prototype.initGrid = /**
     * Initializes the grid that contains
     * the info about the "shape" of the
     * month.
     * @return {?}
     */
    function () {
        /** @type {?} */
        var defaultEmpty = undefined;
        /** @type {?} */
        var grid = [
            new Array(this.DAYS_PER_WEEK).fill(defaultEmpty),
            new Array(this.DAYS_PER_WEEK).fill(defaultEmpty),
            new Array(this.DAYS_PER_WEEK).fill(defaultEmpty),
            new Array(this.DAYS_PER_WEEK).fill(defaultEmpty),
            new Array(this.DAYS_PER_WEEK).fill(defaultEmpty)
        ];
        return grid;
    };
    return MonthView;
}());
/**
 * Creates a view of a month.
 */
export { MonthView };
if (false) {
    /** @type {?} */
    MonthView.prototype.WEEKS_PER_MONTH;
    /** @type {?} */
    MonthView.prototype.DAYS_PER_WEEK;
    /** @type {?} */
    MonthView.prototype.date;
}
