// ngx-simple-calendar/esm5/lib/data-models/data-calendar.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// ENUMS
/** @enum {number} */
var Month = {
    January: 0,
    February: 1,
    March: 2,
    April: 3,
    May: 4,
    June: 5,
    July: 6,
    August: 7,
    September: 8,
    October: 9,
    November: 10,
    December: 11,
};
export { Month };
Month[Month.January] = 'January';
Month[Month.February] = 'February';
Month[Month.March] = 'March';
Month[Month.April] = 'April';
Month[Month.May] = 'May';
Month[Month.June] = 'June';
Month[Month.July] = 'July';
Month[Month.August] = 'August';
Month[Month.September] = 'September';
Month[Month.October] = 'October';
Month[Month.November] = 'November';
Month[Month.December] = 'December';
/** @enum {number} */
var Day = {
    Sunday: 0,
    Monday: 1,
    Tuesday: 2,
    Wednesday: 3,
    Thursday: 4,
    Friday: 5,
    Saturday: 6,
};
export { Day };
Day[Day.Sunday] = 'Sunday';
Day[Day.Monday] = 'Monday';
Day[Day.Tuesday] = 'Tuesday';
Day[Day.Wednesday] = 'Wednesday';
Day[Day.Thursday] = 'Thursday';
Day[Day.Friday] = 'Friday';
Day[Day.Saturday] = 'Saturday';
/**
 * @record
 */
export function CalendarDay() { }
if (false) {
    /** @type {?} */
    CalendarDay.prototype.today;
    /** @type {?} */
    CalendarDay.prototype.number;
    /** @type {?} */
    CalendarDay.prototype.month;
    /** @type {?} */
    CalendarDay.prototype.year;
    /** @type {?} */
    CalendarDay.prototype.startDateTime;
    /** @type {?} */
    CalendarDay.prototype.endDateTime;
    /** @type {?} */
    CalendarDay.prototype.weekend;
    /** @type {?} */
    CalendarDay.prototype.activeMonth;
    /** @type {?} */
    CalendarDay.prototype.events;
}
/**
 * @record
 */
export function TopBar() { }
if (false) {
    /** @type {?} */
    TopBar.prototype.year;
    /** @type {?} */
    TopBar.prototype.month;
    /** @type {?} */
    TopBar.prototype.setNextMonth;
    /** @type {?} */
    TopBar.prototype.setPreviousMonth;
}
/**
 * @record
 */
export function CalendarEvent() { }
if (false) {
    /** @type {?} */
    CalendarEvent.prototype.startDateTime;
    /** @type {?|undefined} */
    CalendarEvent.prototype.endDateTime;
    /** @type {?|undefined} */
    CalendarEvent.prototype.data;
}
/**
 * @record
 */
export function DayI() { }
if (false) {
    /** @type {?} */
    DayI.prototype.id;
    /** @type {?} */
    DayI.prototype.translationKey;
    /** @type {?} */
    DayI.prototype.name;
}
/**
 * @record
 */
export function CalendarMonths() { }
if (false) {
    /** @type {?} */
    CalendarMonths.prototype.id;
    /** @type {?} */
    CalendarMonths.prototype.translationKey;
    /** @type {?} */
    CalendarMonths.prototype.name;
}
// BASE DATA
var ɵ0 = Month.January, ɵ1 = Month.February, ɵ2 = Month.March, ɵ3 = Month.April, ɵ4 = Month.May, ɵ5 = Month.June, ɵ6 = Month.July, ɵ7 = Month.August, ɵ8 = Month.September, ɵ9 = Month.October, ɵ10 = Month.November, ɵ11 = Month.December;
/** @type {?} */
export var months = [{
        id: ɵ0,
        translationKey: 'calendar.months.january',
        name: 'January'
    }, {
        id: ɵ1,
        translationKey: 'calendar.months.february',
        name: 'February'
    }, {
        id: ɵ2,
        translationKey: 'calendar.months.march',
        name: 'March'
    }, {
        id: ɵ3,
        translationKey: 'calendar.months.april',
        name: 'April'
    }, {
        id: ɵ4,
        translationKey: 'calendar.months.may',
        name: 'May'
    }, {
        id: ɵ5,
        translationKey: 'calendar.months.june',
        name: 'June'
    }, {
        id: ɵ6,
        translationKey: 'calendar.months.july',
        name: 'July'
    }, {
        id: ɵ7,
        translationKey: 'calendar.months.august',
        name: 'August'
    }, {
        id: ɵ8,
        translationKey: 'calendar.months.september',
        name: 'September'
    }, {
        id: ɵ9,
        translationKey: 'calendar.months.october',
        name: 'October'
    }, {
        id: ɵ10,
        translationKey: 'calendar.months.november',
        name: 'November'
    }, {
        id: ɵ11,
        translationKey: 'calendar.months.december',
        name: 'December'
    }];
var ɵ12 = Day.Monday, ɵ13 = Day.Tuesday, ɵ14 = Day.Wednesday, ɵ15 = Day.Thursday, ɵ16 = Day.Friday, ɵ17 = Day.Saturday, ɵ18 = Day.Sunday;
/** @type {?} */
export var days = [{
        id: ɵ12,
        translationKey: 'calendar.days.monday',
        name: 'Monday'
    }, {
        id: ɵ13,
        translationKey: 'calendar.days.tuesday',
        name: 'Tuesday'
    }, {
        id: ɵ14,
        translationKey: 'calendar.days.wednesday',
        name: 'Wednesday'
    }, {
        id: ɵ15,
        translationKey: 'calendar.days.thursday',
        name: 'Thursday'
    }, {
        id: ɵ16,
        translationKey: 'calendar.days.friday',
        name: 'Friday'
    }, {
        id: ɵ17,
        translationKey: 'calendar.days.saturday',
        name: 'Saturday'
    }, {
        id: ɵ18,
        translationKey: 'calendar.days.sunday',
        name: 'Sunday'
    }];
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9, ɵ10, ɵ11, ɵ12, ɵ13, ɵ14, ɵ15, ɵ16, ɵ17, ɵ18 };
